[WIP] Migrate Configure > Advanced Parameters > Webservices

matks opened the pull request:
Linked to https://github.com/PrestaShop/prestafony-project/issues/37

⚠️  This is a WIP page as I need help from @mickaelandrieu for the webservice keys listing features.

<hr>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop |
| Description?  | Migrate BO page 'webservice' to Symfony using modern controllers and form handlers |
| Type?         | new feature |
| Category?     | BO |
| BC breaks?    | yes |
| Deprecations? | no |
| Fixed ticket? |  |
| How to test?  | See below |

### How to test
Login on backoffice, then navigate to "Webservice page". For example use left menu "Configure > Advanced Parameters > Webservice" link.

Check that:
- the webservice keys listing is working (sorting, pagination, filtering, search, bulk actions, link)
- the "Configuration" form to enable Webservice usage and CGI mode is working
- the "add a new / edit webservice key" page is working (you can access it using "add a new key" button or a "edit" button on one webservice key)

#### BC break
As the Controller and the Template are being moved, this will disable any override of the previous controller and template.

### Possible issues because of the migration

> URLs have changed

Previous URL:
- `/admin/index.php?controller=AdminWebservice`
New URLs:
- `/admin/configure/advanced/webservice/add`
- `/admin/configure/advanced/webservice/`


> Design have changed

The new page is powered by Symfony and uses the new admin theme. I tried to make it look as much as possible to the previous one but the design is slightly different (example: previous theme used bootstrap icons, new theme uses material icons).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/9272)
<!-- Reviewable:end -->

matks committed:
Migrate BO Advanced > Webservice 'enable configuration' form to SF

matks committed:
Enable 1st version of webservice keys listing with grid component for Admin Webservice page

matks committed:
Add unit test for service WebserviceCanBeEnabledConfigurationChecker

matks committed:
Allows 2 forms handling on Admin Webservice page

matks commented:
@mickaelandrieu This is what the page looks like:

![capture d ecran 2018-07-06 a 11 29 28](https://user-images.githubusercontent.com/3830050/42371497-ec3a2994-810f-11e8-88d8-5d227a593d80.png)

mickaelandrieu commented on src/Core/Grid/Definition/Factory/WebserviceKeyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,115 @@
+            ;
+    }
+
+    /**
```
by default they are null I guess, so you don't need to call them at all, right?

matks commented:
@mickaelandrieu On the webservice keys listing
- search in text works (👍 )
- sorting a column does not work
- bulk actions do not work
- global actions ("add new", "refresh list", "show sql", "export to sql manager") dot not work
- row action ("edit" and "delete" one row) do not work

I tried to enable these features by looking at internal behavior of the Logs page but there is a mix of JS to generate, generic Twig templating, specific-to-logs-page Twig templating, options in the Symfony Form that I could not sort out.

I need your guidance 🙏 for these.

I think I can figure out the rest myself (help button link, pagination, drop-down filter menu for column "Enabled").

mickaelandrieu commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/WebserviceController.php:
```diff
@@ -0,0 +1,180 @@
+         * - foreach item: edit/delete
+         */
+
+        if('POST' === $request->getMethod()) {
```
you can use $request->isMethod('POST')

sarjon replied:
or even `$request->isMethod(Request::METHOD_POST)`.

mickaelandrieu commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/WebserviceController.php:
```diff
@@ -0,0 +1,180 @@
+     */
+    private function handleListingSearch(Request $request)
+    {
+        // temporary search criteria class, to be removed ? see TemporarySearchCriteria
```
yes, for information you're stuck waiting for thisto be merged  => https://github.com/PrestaShop/PrestaShop/pull/8984

mickaelandrieu commented on src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/enabled_or_not.html.twig:
```diff
@@ -0,0 +1,30 @@
+    <i class="material-icons">done</i>
+{% else %}
+    <i class="material-icons">not_interested</i>
+{% endif %}
```
EOL

mickaelandrieu commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/WebservicePage/webservice.html.twig:
```diff
@@ -0,0 +1,82 @@
+    </div>
+    {% endblock %}
+
+{% endblock %}
```
EOL

mickaelandrieu reviewed (changes requested):
it's only a first review.

Good job @matks !

mickaelandrieu commented:
>  sorting a column does not work

it's fixed by https://github.com/PrestaShop/PrestaShop/pull/8984

> bulk actions do not work

It's because you didn't enable the Grid in javascript, see => https://github.com/PrestaShop/PrestaShop/pull/8984/files#diff-2c7b6c70ffc40819eaa238309d19d0a6R31 (and don't forget to import grid library). Not your fault, it's not documented :(

> global actions ("add new", "refresh list", "show sql", "export to sql manager") dot not work

Same as before

> row action ("edit" and "delete" one row) do not work

Did you map them to a Symfony action? This is not magic, you need to declare both "edit" and "delete" actions and map them to controller actions. We already know at this moment these 2 actions can be abstracted, don't you think? Can we make something re-usable?

mickaelandrieu commented:
ping @sarjon would you mind to review? :+1:

sarjon commented on src/Adapter/Webservice/WebserviceConfiguration.php:
```diff
@@ -0,0 +1,83 @@
+/**
+ * Manages the configuration data about upload quota options.
+ */
+class WebserviceConfiguration implements DataConfigurationInterface
```
you can make it `final`.

sarjon commented on src/Adapter/Webservice/WebserviceConfiguration.php:
```diff
@@ -0,0 +1,83 @@
+     */
+    private $configuration;
+
+    public function __construct(Configuration $configuration)
```
missing docblocks.

sarjon commented on src/Core/Grid/Column/Type/Common/EnabledOrNotColumn.php:
```diff
@@ -0,0 +1,19 @@
+<?php
```
missing license header.

sarjon commented on src/Core/Grid/Column/Type/Common/EnabledOrNotColumn.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Column used to display booleans as icons for "enabled" or "active" properties
+ */
+final class EnabledOrNotColumn extends AbstractColumn
```
hmm, maybe we can name it `StatusColumn` or similar? :/

matks replied:
In fact I realized it should be `ToggleColumn` because in the legacy page, this column allows to toggle the value of the boolean property `active`being displayed

sarjon commented on src/Core/Grid/Search/WebserviceKeyGridSearchCriteria.php:
```diff
@@ -0,0 +1,68 @@
+
+use Symfony\Component\HttpFoundation\Request;
+
+class WebserviceKeyGridSearchCriteria implements SearchCriteriaInterface
```
should make it `final`. 

missing docblock and same for all methods in the class.

sarjon commented on src/Core/Webservice/WebserviceCanBeEnabledConfigurationChecker.php:
```diff
@@ -0,0 +1,162 @@
+        $this->configuration = $configuration;
+    }
+
+    public function analyseConfigurationForIssues(Request $request = null)
```
Please describe (add docblock) what methods are suppose to do. :)

sarjon commented on src/Core/Webservice/WebserviceCanBeEnabledConfigurationChecker.php:
```diff
@@ -0,0 +1,162 @@
+    {
+        $issues = [];
+
+        if ($request !== null) {
```
i believe you can merge it into single `if`.

`if ($request !== null && false === strpos($request->server->get('SERVER_SOFTWARE'), 'Apache'))` ?

or even give it a context for better readability, like:
```php
$isUsingApacheWebserver = /* your condition */;
if (!$isUsingApacheWebserver) {
    // ...
}
`

sarjon commented on src/Core/Webservice/WebserviceCanBeEnabledConfigurationChecker.php:
```diff
@@ -0,0 +1,162 @@
+/**
+ * Looks at server configuration in order to check PrestaShop Webservice can be enabled
+ */
+class WebserviceCanBeEnabledConfigurationChecker
```
you can introduce new interface for it and make it final. :+1:

mickaelandrieu replied:
not sure this need to be interfaced as this class is already the contract if we make it final.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/WebserviceController.php:
```diff
@@ -0,0 +1,180 @@
+ */
+class WebserviceController extends FrameworkBundleAdminController
+{
+    const CONTROLLER_NAME = 'AdminWebservice';
```
this constant is not needed, you can access legacy controller name by
`$request->attributes->get('_legacy_controller')`.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/WebserviceController.php:
```diff
@@ -0,0 +1,180 @@
+            'requireBulkActions' => false, // temporary
+            'showContentHeader' => true,
+            'enableSidebar' => true,
+            'help_link' => $this->generateSidebarLink('AdminWebservice'),
```
remove hardcoded `AdminWebservice`.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/WebserviceController.php:
```diff
@@ -0,0 +1,180 @@
+     * @param FormInterface $form
+     * @return Response
+     */
+    public function indexAction(FormInterface $form = null, Request $request)
```
is there a case when `FormInterface` is actually injected?

sarjon commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/WebserviceController.php:
```diff
@@ -0,0 +1,180 @@
+        ];
+
+        return $this->render('@AdvancedParameters/WebservicePage/webservice.html.twig', $twigValues);
+
```
remove empty line.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/WebserviceController.php:
```diff
@@ -0,0 +1,180 @@
+
+        $configurationWarnings = $this->lookForWarnings($request);
+
+        if (false === empty($configurationWarnings)) {
```
```diff
- false === empty($configurationWarnings)
+ !empty($configurationWarnings)
```

sarjon commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/WebserviceController.php:
```diff
@@ -0,0 +1,180 @@
+        if ($form->isSubmitted()) {
+            $data = $form->getData();
+
+            $saveErrors = $this->getFormHandler()->save($data);
```
you can use `$form->getData()` directly

sarjon commented on src/PrestaShopBundle/Entity/Repository/WebserviceKeyRepository.php:
```diff
@@ -0,0 +1,145 @@
+class WebserviceKeyRepository implements RepositoryInterface, DoctrineQueryBuilderInterface
+{
+    /** @var \Doctrine\DBAL\Driver\Connection */
+    private $connection;
```
please update formatting, like:

```php
/**
 * @var \Doctrine\DBAL\Driver\Connection
 */
private $connection;
```

sarjon commented on src/PrestaShopBundle/Entity/Repository/WebserviceKeyRepository.php:
```diff
@@ -0,0 +1,145 @@
+    /** @var string */
+    private $webserviceKeyTable;
+
+    public function __construct(Connection $connection, $databasePrefix)
```
missing dockblock

sarjon commented on src/PrestaShopBundle/Entity/Repository/WebserviceKeyRepository.php:
```diff
@@ -0,0 +1,145 @@
+
+        $filters = $searchCriteria->getFilters();
+        //var_dump($searchCriteria);
+        //die();
```
dont forget to remove these. :)

sarjon reviewed (commented):
@mickaelandrieu  quick review done :+1:

matks committed:
Allow Grid row items actions: edit and delete for Admin Webservice page

matks committed:
Clean code for Admin Webservice page

matks commented:
@sarjon @mickaelandrieu I cleaned my code following your reviews and I introduced actions "edit" and "delete" for each of my grid rows in https://github.com/PrestaShop/PrestaShop/pull/9272/commits/5eda6736c45bc42ffd95e233f60b94fe7d39334e and https://github.com/PrestaShop/PrestaShop/pull/9272/commits/95e07c56aa447a2d038bae78983038cb235726a0.

I introduced an [ActionButton](https://github.com/PrestaShop/PrestaShop/pull/9272/commits/5eda6736c45bc42ffd95e233f60b94fe7d39334e#diff-37e6bf9f962212db55f516a762d5bd0e) to be given to  an ActionColumn. I did it because I looked at [action.html.twig](https://github.com/PrestaShop/PrestaShop/pull/9272/commits/5eda6736c45bc42ffd95e233f60b94fe7d39334e#diff-e0f6fa6c08b185b2f0155dc15621f7ebL26) and I saw a "@todo".

However after looking deeper at your code, it looks like I might have duplicated the behavior from [RowAction](https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Action/RowAction.php). Can you confirm ?

sarjon commented:
> However after looking deeper at your code, it looks like I might have duplicated the behavior from RowAction. Can you confirm ?

yes, actually `RowAction` was implemented for this purpose.

i've implemented same functionality on another page, you can check it's PR here: https://github.com/PrestaShop/PrestaShop/pull/9206/files#diff-724b095480c951f1faf90922d0dd4dcaR87

and this is how it looks: https://prnt.sc/k3f5uf

@mickaelandrieu i think @matks is working on something i've already implemented in SQL manager PR migration. :/

mickaelandrieu commented:
@matks please reuse what is done by @sarjon for this specific issue, I try to make his contribution merged asap

LouiseBonnard commented on src/Core/Grid/Definition/Factory/WebserviceKeyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,154 @@
+     */
+    protected function getName()
+    {
+        return $this->trans('Webservice keys', [], 'Admin.Navigation.Menu');
```
Right string is 'Webservice' in _Admin.Navigation.Menu_

LouiseBonnard commented on src/Core/Grid/Definition/Factory/WebserviceKeyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,154 @@
+                ->setName($this->trans('Key', [], 'Global.Actions'))
+            )
+            ->add((new SimpleColumn('description'))
+                ->setName($this->trans('Key description', [], 'Global.Actions'))
```
Wrong domains: _Admin.Advparameters.Feature_

LouiseBonnard commented on src/Core/Grid/Definition/Factory/WebserviceKeyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,154 @@
+                ])
+            )
+            ->add((new SimpleColumn('key'))
+                ->setName($this->trans('Key', [], 'Global.Actions'))
```
Wrong domains: _Admin.Advparameters.Feature_

LouiseBonnard commented on src/Core/Grid/Definition/Factory/WebserviceKeyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,154 @@
+                ->setName($this->trans('Key description', [], 'Global.Actions'))
+            )
+            ->add((new EnabledOrNotColumn('active')) // @todo: this should be a ToggleColumn instead
+                ->setName($this->trans('Enabled', [], 'Global.Actions'))
```
Wrong domains: _Admin.Global_

LouiseBonnard commented on src/Core/Grid/Definition/Factory/WebserviceKeyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,154 @@
+                ])
+            )
+            ->add((new ActionColumn('actions'))
+                ->setName($this->trans('Actions', [], 'Global.Actions'))
```
Wrong domains: _Admin.Global_

LouiseBonnard commented on src/Core/Grid/Definition/Factory/WebserviceKeyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,154 @@
+                ->setOptions([
+                    'filter_type' => SubmitType::class,
+                    'filter_type_options' => [
+                        'label' => $this->trans('Search', [], 'Global.Actions'),
```
Wrong domains: _Admin.Actions_

PS: there is no 'Global' domain, only four main repertories (Admin, Emails, Modules and Shop)

LouiseBonnard commented on src/Core/Grid/Definition/Factory/WebserviceKeyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,154 @@
+                            $this->router,
+                            'admin_webservice_edit_key',
+                            ['webservice_key_id' => 'id_webservice_account'], // should it be an object ? like a Mapping or ParameterMapping ?
+                            $this->trans('Edit', [], 'Global.Actions'),
```
Wrong domains: _Admin.Actions_

LouiseBonnard commented on src/Core/Grid/Definition/Factory/WebserviceKeyGridDefinitionFactory.php:
```diff
@@ -0,0 +1,154 @@
+                            $this->router,
+                            'admin_webservice_delete_key',
+                            ['webservice_key_id' => 'id_webservice_account'], // should it be an object ? like a Mapping or ParameterMapping ?
+                            $this->trans('Delete', [], 'Global.Actions'),
```
Wrong domains: _Admin.Actions_

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/WebservicePage/webservice.html.twig:
```diff
@@ -0,0 +1,81 @@
+                <div class="col">
+                    <div class="card">
+                        <h3 class="card-header">
+                            <i class="material-icons">settings</i> {{ 'Configuration'|trans }}
```
Wrong domains: _Admin.Global_

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/WebservicePage/webservice.html.twig:
```diff
@@ -0,0 +1,81 @@
+                                    {{ form_widget(form.webservice_configuration.enable_webservice) }}
+                                </div>
+
+                                <small class="form-text">{{ 'Before activating the webservice, you must be sure to:\n 1. Check that URL rewriting is available on this server.\n 2.Check that the five methods GET, POST, PUT, DELETE and HEAD are supported by this server.'|trans({}, 'Admin.Shopparameters.Help')|nl2br }}</small>
```
Is the whole string wrote properly? I have three separate strings in Crowdin.

LouiseBonnard replied:
In _Admin.Advparameters.Help_ !

mickaelandrieu replied:
you're [right](https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminWebserviceController.php#L81) @LouiseBonnard.

Please @matks be careful with translation keys, if we change them we lost all related translations.

matks replied:
@mickaelandrieu Note to myself: add this to the Migration Guide

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/WebservicePage/webservice.html.twig:
```diff
@@ -0,0 +1,81 @@
+                                    {{ form_widget(form.webservice_configuration.enable_cgi) }}
+                                </div>
+
+                                <small class="form-text">{{ 'Before choosing "Yes", check that PHP is not configured as an Apache module on your server.'|trans({}, 'Admin.Shopparameters.Help') }}</small>
```
Wrong domains: _Admin.Advparameters.Help_

mickaelandrieu commented:
Thanks for translations review @LouiseBonnard :)

LouiseBonnard commented:
My pleasure. ;-)

matks commented:
This PR is too big, it will be split.
Part 1 is PR https://github.com/PrestaShop/PrestaShop/pull/9310. Parts 2 and 3 (webservice keys listing and "add/edit a webservice key" form) will come later.
